### PR TITLE
[RFR] Fix optimistic delete in List

### PR DESCRIPTION
**Pull request by djhi**

### CODE REVIEW

File: packages/ra-core/src/reducer/admin/resource/data.js [hunk 1]
```diff
@@ -69,6 +71,42 @@ export default (previousState = initialState, { type, payload, meta }) => {
             .map(record => ({ ...record, ...payload.data }));
         return addRecords(updatedRecords, previousState);
     }
+    if (type === CRUD_DELETE_OPTIMISTIC) {
+        const newState = Object.keys(previousState).reduce((acc, id) => {
```
Comment (fzaninotto) [inline packages/ra-core/src/reducer/admin/resource/data.js]: You're doing an `omit` here, which can be written in one line in ES6:

```js
const { id, ...nextState } = previousState;
```

File: packages/ra-core/src/reducer/admin/resource/data.js [hunk 2]
```diff
@@ -69,6 +71,42 @@ export default (previousState = initialState, { type, payload, meta }) => {
...
+        }, {});
+
+        Object.defineProperty(newState, 'fetchedAt', {
+            value: previousState.fetchedAt,
+        });
+
+        return newState;
+    }
+    if (type === CRUD_DELETE_MANY_OPTIMISTIC) {
+        const newState = Object.keys(previousState).reduce((acc, id) => {
```
Comment (fzaninotto) [inline packages/ra-core/src/reducer/admin/resource/data.js]: For clarity, I'd rewrite this one with a `filter` and a simpler `reduce`, as follows:

```js
const newState = Object.entries(previousState)
   .filter(([key]) => !payload.ids.includes(key))
   .reduce((obj, [key, val]) => ({ ...obj, [key]: val }), {});
```

File: packages/ra-core/src/reducer/admin/resource/data.js (hunk 1 above)
Comment (djhi) [reply packages/ra-core/src/reducer/admin/resource/data.js]: mmm `id` is a variable, not a string key

Comment (djhi) [reply packages/ra-core/src/reducer/admin/resource/data.js]: `const { [payload.id]: removed, ...newState } = previousState;` :tada:

Comment (djhi) [reply packages/ra-core/src/reducer/admin/resource/data.js]: I :heart: es6

**Comment by fzaninotto** [comment]
Would you mind adding a little unit test at the reducer level to prove that your fix works?

File: packages/ra-core/src/reducer/admin/resource/data.js (hunk 1 above)
Comment (fzaninotto) [reply packages/ra-core/src/reducer/admin/resource/data.js]: 💪 much better

**Pull request merged**